Error on document close, after renaming a file

In vscode 1.32.3 (1.33 has the same problem but has [issues of its own](https://github.com/Microsoft/vscode/issues/72343)), after renaming a type, when you save the file (which has been renamed) the server throws an exception in JDTLanguageServer.didClose:

```
[Trace - 9:57:49 AM] Sending request 'textDocument/rename - (12)'.
Params: {
    "textDocument": {
        "uri": "file:///Users/fbricon/Dev/souk/getting-quarked/src/main/java/org/acme/quickstart/Bar.java"
    },
    "position": {
        "line": 5,
        "character": 14
    },
    "newName": "Fooo"
}


[Trace - 9:57:49 AM] Received response 'textDocument/rename - (12)' in 194ms.
Result: {
    "changes": {},
    "documentChanges": [
        {
            "textDocument": {
                "version": 0,
                "uri": "file:///Users/fbricon/Dev/souk/getting-quarked/src/main/java/org/acme/quickstart/Bar.java"
            },
            "edits": [
                {
                    "range": {
                        "start": {
                            "line": 5,
                            "character": 13
                        },
                        "end": {
                            "line": 5,
                            "character": 16
                        }
                    },
                    "newText": "Fooo"
                }
            ]
        },
        {
            "oldUri": "file:///Users/fbricon/Dev/souk/getting-quarked/src/main/java/org/acme/quickstart/Bar.java",
            "newUri": "file:///Users/fbricon/Dev/souk/getting-quarked/src/main/java/org/acme/quickstart/Fooo.java",
            "kind": "rename"
        }
    ]
}


[Trace - 9:57:49 AM] Sending notification 'textDocument/didChange'.
Params: {
    "textDocument": {
        "uri": "file:///Users/fbricon/Dev/souk/getting-quarked/src/main/java/org/acme/quickstart/Bar.java",
        "version": 2
    },
    "contentChanges": [
        {
            "range": {
                "start": {
                    "line": 5,
                    "character": 13
                },
                "end": {
                    "line": 5,
                    "character": 16
                }
            },
            "rangeLength": 3,
            "text": "Fooo"
        }
    ]
}


[Trace - 9:57:49 AM] Sending notification 'textDocument/didClose'.
Params: {
    "textDocument": {
        "uri": "file:///Users/fbricon/Dev/souk/getting-quarked/src/main/java/org/acme/quickstart/Bar.java"
    }
}


[Trace - 9:57:49 AM] Received notification 'window/logMessage'.
Params: {
    "type": 1,
    "message": "Apr. 15, 2019, 9:57:49 a.m. Error while handling document close. URI: file:///Users/fbricon/Dev/souk/getting-quarked/src/main/java/org/acme/quickstart/Bar.java\nFile not found: /Users/fbricon/Dev/souk/getting-quarked/src/main/java/org/acme/quickstart/Bar.java.\nJava Model Exception: Core Exception [code 368] File not found: /Users/fbricon/Dev/souk/getting-quarked/src/main/java/org/acme/quickstart/Bar.java....

[Error - 9:57:49 AM] Apr. 15, 2019, 9:57:49 a.m. Error while handling document close. URI: file:///Users/fbricon/Dev/souk/getting-quarked/src/main/java/org/acme/quickstart/Bar.java
File not found: /Users/fbricon/Dev/souk/getting-quarked/src/main/java/org/acme/quickstart/Bar.java.
Java Model Exception: Core Exception [code 368] File not found: /Users/fbricon/Dev/souk/getting-quarked/src/main/java/org/acme/quickstart/Bar.java.
	at org.eclipse.jdt.internal.core.util.Util.getResourceContentsAsCharArray(Util.java:1209)
	at org.eclipse.jdt.internal.core.util.Util.getResourceContentsAsCharArray(Util.java:1181)
	at org.eclipse.jdt.internal.core.CompilationUnit.openBuffer(CompilationUnit.java:1215)
	at org.eclipse.jdt.internal.core.CompilationUnit.buildStructure(CompilationUnit.java:115)
	at org.eclipse.jdt.internal.core.Openable.generateInfos(Openable.java:268)
	at org.eclipse.jdt.internal.core.JavaElement.openWhenClosed(JavaElement.java:596)
	at org.eclipse.jdt.internal.core.BecomeWorkingCopyOperation.executeOperation(BecomeWorkingCopyOperation.java:43)
	at org.eclipse.jdt.internal.core.JavaModelOperation.run(JavaModelOperation.java:736)
	at org.eclipse.jdt.internal.core.JavaModelOperation.runOperation(JavaModelOperation.java:802)
	at org.eclipse.jdt.internal.core.CompilationUnit.becomeWorkingCopy(CompilationUnit.java:99)
	at org.eclipse.jdt.internal.core.CompilationUnit.becomeWorkingCopy(CompilationUnit.java:106)
	at org.eclipse.jdt.ls.core.internal.handlers.DocumentLifeCycleHandler.handleClosed(DocumentLifeCycleHandler.java:392)
	at org.eclipse.jdt.ls.core.internal.handlers.DocumentLifeCycleHandler$3.run(DocumentLifeCycleHandler.java:200)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2295)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2317)
	at org.eclipse.jdt.ls.core.internal.handlers.DocumentLifeCycleHandler.didClose(DocumentLifeCycleHandler.java:197)
	at org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer.didClose(JDTLanguageServer.java:693)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:65)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.notify(GenericEndpoint.java:152)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleNotification(RemoteEndpoint.java:220)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:187)
	at org.eclipse.jdt.ls.core.internal.ParentProcessWatcher.lambda$0(ParentProcessWatcher.java:83)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:192)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:99)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: org.eclipse.core.internal.resources.ResourceException: File not found: /Users/fbricon/Dev/souk/getting-quarked/src/main/java/org/acme/quickstart/Bar.java.
	at org.eclipse.core.internal.localstore.FileSystemResourceManager.read(FileSystemResourceManager.java:833)
	at org.eclipse.core.internal.resources.File.getContents(File.java:280)
	at org.eclipse.jdt.internal.core.util.Util.getResourceContentsAsCharArray(Util.java:1207)
	... 33 more
Caused by: java.lang.Exception: File not found: /Users/fbricon/Dev/souk/getting-quarked/src/main/java/org/acme/quickstart/Bar.java.
	at org.eclipse.core.internal.resources.ResourceException.provideStackTrace(ResourceException.java:42)
	at org.eclipse.core.internal.resources.ResourceException.<init>(ResourceException.java:38)
	... 36 more
Caused by: org.eclipse.core.internal.resources.ResourceException(/getting-quarked/src/main/java/org/acme/quickstart/Bar.java)[368]: java.lang.Exception: File not found: /Users/fbricon/Dev/souk/getting-quarked/src/main/java/org/acme/quickstart/Bar.java.
	at org.eclipse.core.internal.resources.ResourceException.provideStackTrace(ResourceException.java:42)
	at org.eclipse.core.internal.resources.ResourceException.<init>(ResourceException.java:38)
	at org.eclipse.core.internal.localstore.FileSystemResourceManager.read(FileSystemResourceManager.java:833)
	at org.eclipse.core.internal.resources.File.getContents(File.java:280)
	at org.eclipse.jdt.internal.core.util.Util.getResourceContentsAsCharArray(Util.java:1207)
	at org.eclipse.jdt.internal.core.util.Util.getResourceContentsAsCharArray(Util.java:1181)
	at org.eclipse.jdt.internal.core.CompilationUnit.openBuffer(CompilationUnit.java:1215)
	at org.eclipse.jdt.internal.core.CompilationUnit.buildStructure(CompilationUnit.java:115)
	at org.eclipse.jdt.internal.core.Openable.generateInfos(Openable.java:268)
	at org.eclipse.jdt.internal.core.JavaElement.openWhenClosed(JavaElement.java:596)
	at org.eclipse.jdt.internal.core.BecomeWorkingCopyOperation.executeOperation(BecomeWorkingCopyOperation.java:43)
	at org.eclipse.jdt.internal.core.JavaModelOperation.run(JavaModelOperation.java:736)
	at org.eclipse.jdt.internal.core.JavaModelOperation.runOperation(JavaModelOperation.java:802)
	at org.eclipse.jdt.internal.core.CompilationUnit.becomeWorkingCopy(CompilationUnit.java:99)
	at org.eclipse.jdt.internal.core.CompilationUnit.becomeWorkingCopy(CompilationUnit.java:106)
	at org.eclipse.jdt.ls.core.internal.handlers.DocumentLifeCycleHandler.handleClosed(DocumentLifeCycleHandler.java:392)
	at org.eclipse.jdt.ls.core.internal.handlers.DocumentLifeCycleHandler$3.run(DocumentLifeCycleHandler.java:200)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2295)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2317)
	at org.eclipse.jdt.ls.core.internal.handlers.DocumentLifeCycleHandler.didClose(DocumentLifeCycleHandler.java:197)
	at org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer.didClose(JDTLanguageServer.java:693)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:65)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.notify(GenericEndpoint.java:152)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleNotification(RemoteEndpoint.java:220)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:187)
	at org.eclipse.jdt.ls.core.internal.ParentProcessWatcher.lambda$0(ParentProcessWatcher.java:83)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:192)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:99)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)

[Trace - 9:57:49 AM] Sending notification 'textDocument/didOpen'.
Params: {
    "textDocument": {
        "uri": "file:///Users/fbricon/Dev/souk/getting-quarked/src/main/java/org/acme/quickstart/Fooo.java",
        "languageId": "java",
        "version": 1,
        "text": "package org.acme.quickstart;\n\n/**\n * Foo\n */\npublic class Fooo {\n\n    \n}"
    }
}

}
```

--- a/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/DocumentLifeCycleHandler.java
+++ b/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/DocumentLifeCycleHandler.java
@@ -439,7 +439,7 @@ public class DocumentLifeCycleHandler {
 			}
 			if (JDTUtils.isDefaultProject(unit) || !JDTUtils.isOnClassPath(unit) || unit.getResource().isDerived()) {
 				new DiagnosticsHandler(connection, unit).clearDiagnostics();
-			} else if (unit.hasUnsavedChanges()) {
+			} else if (hasUnsavedChanges(unit)) {
 				unit.discardWorkingCopy();
 				unit.becomeWorkingCopy(new NullProgressMonitor());
 				publishDiagnostics(unit, new NullProgressMonitor());
@@ -458,6 +458,14 @@ public class DocumentLifeCycleHandler {
 		} catch (CoreException e) {
 			JavaLanguageServerPlugin.logException("Error while handling document close. URI: " + uri, e);
 		}
+	}
+
+	private boolean hasUnsavedChanges(ICompilationUnit unit) throws CoreException {
+		if (!unit.hasUnsavedChanges()) {
+			return false;
+		}
+		unit.getResource().refreshLocal(IResource.DEPTH_ZERO, null);
+		return unit.getResource().exists();
 	}
 
 	public void handleSaved(DidSaveTextDocumentParams params) {

--- a/org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/handlers/DocumentLifeCycleHandlerTest.java
+++ b/org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/handlers/DocumentLifeCycleHandlerTest.java
@@ -13,6 +13,7 @@ package org.eclipse.jdt.ls.core.internal.handlers;
 import static org.eclipse.jdt.ls.core.internal.Lsp4jAssertions.assertRange;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
 import static org.mockito.Mockito.when;
 
 import java.io.File;
@@ -673,6 +674,49 @@ public class DocumentLifeCycleHandlerTest extends AbstractProjectsManagerBasedTe
 		astRoot = CoreASTProvider.getInstance().getAST(cu, CoreASTProvider.WAIT_YES, new NullProgressMonitor());
 		problems = astRoot.getProblems();
 		assertEquals("Unexpected number of errors", 1, problems.length);
+	}
+
+	@Test
+	public void testCloseMissingResource() throws Exception {
+		IJavaProject javaProject = newEmptyProject();
+		IPackageFragmentRoot sourceFolder = javaProject.getPackageFragmentRoot(javaProject.getProject().getFolder("src"));
+		IPackageFragment pack1 = sourceFolder.createPackageFragment("test1", false, null);
+
+		StringBuilder buf = new StringBuilder();
+		buf.append("package test1;\n");
+		buf.append("public class E123 {\n");
+		buf.append("    public boolean foo() {\n");
+		buf.append("        return x;\n");
+		buf.append("    }\n");
+		buf.append("}\n");
+		ICompilationUnit cu1 = pack1.createCompilationUnit("E123.java", buf.toString(), false, null);
+
+		openDocument(cu1, cu1.getSource(), 1);
+
+		assertEquals(true, cu1.isWorkingCopy());
+		assertEquals(false, cu1.hasUnsavedChanges());
+		assertNewProblemReported(new ExpectedProblemReport(cu1, 1));
+		assertEquals(1, getCacheSize());
+		assertNewASTsCreated(1);
+
+		StringBuilder buf2 = new StringBuilder();
+		buf2.append("package test1;\n");
+		buf2.append("public class E123 {\n");
+		buf2.append("    public boolean foo() {\n");
+		buf2.append("        return true;\n");
+		buf2.append("    }\n");
+		buf2.append("}\n");
+		changeDocumentFull(cu1, buf2.toString(), 2);
+		File file = cu1.getResource().getRawLocation().toFile();
+		boolean deleted = file.delete();
+		assertTrue(file.getAbsolutePath() + " hasn't been deleted", deleted);
+		closeDocument(cu1);
+
+		assertEquals(false, cu1.isWorkingCopy());
+		assertEquals(false, cu1.hasUnsavedChanges());
+		assertNewProblemReported(new ExpectedProblemReport(cu1, 0));
+		assertEquals(0, getCacheSize());
+		assertNewASTsCreated(0);
 	}
 
 	private File createTempFile(File parent, String fileName, String content) throws IOException {

